aip/general/0236.md

```diff
@@ -0,0 +1,355 @@
+
+## System Changes
+
+The API logic of this mechanism pushes an “experimental” version of a proposed change to the downstream system (data plane) while the existing “effective” version remains active. Both versions of the configuration are subsequently evaluated for incoming traffic, however, only the effective version’s configurations actually have the ability to alter traffic. Meanwhile, the output of the experimental version is logged so a customer can validate that the proposed changes are as intended. This gives the customer the ability to fully understand the effects of the new configuration against live traffic before replacing the effective configuration with it.
+
+## Non-Goals
+
+### Non-Policy Resources
+
+This rollout safety mechanism is for policies only.  
+
+A policy is a customer-owned resource with a set of rules that govern the behavior of how GCP resources admit or deny a target entity (e.g. network traffic for firewall rules, general GCP resources with org policy). Generally, the outcome of a policy can be evaluated to a specific set of outcomes, for example, FirewallPolicy evaluates to either ALLOW or DENY for incoming traffic into a Virtual Network. This lends well to preview mode as a rollout safety mechanism as the proposed changes to a policy can simply be logged instead of enforced.
+
+### Other Rollout Safety Mechanisms
+
+This AIP does not contain guidance for the following rollout safety mechanisms which are commonly grouped together with preview mode.
+
+- Simulation: Replay old production traffic on the new version of a resource and evaluate the results using logging.
+- Offline-validation: Evaluate new version of a resource using a set of user-defined results like CAS/presubmit checks.
+
+## Guidance
+
+### Supported Use-Cases
+
+The following use-cases are considered when making a recommendation for API best practices of preview mode:
+- Resource: Each entity to be previewed is a REST resource.
+- Collection: A collection of objects live on a REST resource and one or many of these can be previewed.
+Each API recommended below will have a table explaining how it translates to each use-case.
+
+### Nested Collection called Experiments
+
+A new configuration of a policy which is to be validated using preview mode is stored in a nested collection under that policy resource.
+
+In this AIP, we will use the example of a hypothetical policy resource called “Policy”.
+
+- /compute/v1/locations/global/policies/{policy_id}
+
+The experimental versions of the resource used for preview mode are stored as a nested collection underneath the policy. The resource type will follow the convention "<resource_name> + Experiment”.
+
+- /compute/v1/locations/global/policies/{policy_id}/policyExperiments/{policy_experiment_id}
```

Comment:
Nested resources which repeat the parent resource name in the message should not also repeat the parent name in the collection constant and id parameter. Consider `.../experiments/{experiment_id}` as the collection and parameter of the PolicyExperiment.

Policy
.../policies/{policy_id}

PolicyExperiment
.../policies/{policy_id}/experiments/{experiment_id}

Same rules that apply to {Resource}Revision, really.

Book
.../books/{book_id}

BookRevision
.../books/{book_id}/revisions/{revision_id}

Reply:
@toumorokoshi and I aligned on this being standard to have the name of the collection be equal to the name of the resource type. Can you two resolve this and come up with an answer?

Reply:
I have changed this to experiments now to align with the new guidance.